src/covid_model_seiir_pipeline/marshall.py

```diff
@@ -0,0 +1,451 @@
+    date = "dates"
+    fit_beta = "betas"  # TODO: no longer needed - fit is not a stage
+    parameter = "parameters"
+    regression_beta = "beta"
+
+    # types which serialize to a DataFrame
+    DataFrame_types = frozenset([
+        fit_beta,
+        parameter,
+        date,
+        coefficient,
+        regression_beta,
+        beta_scales,
+        components,
+    ])
+
+
+class Keys:
+    """
+    Identifies data payload types and stores related information.
+
+    Callers should never use init and instead use one of the many factory
+    methods which embed important related information to the variable
+    draw_id/location_id argument.
+    """
+    def __init__(self, data_type, template, **key_args):
+        self.data_type = data_type
+        self.template = template
+        self.key_args = key_args
+
+    # factories related to regression
+    @classmethod
+    def coefficient(cls, draw_id):
+        return cls(DataTypes.coefficient, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def date(cls, draw_id):
+        return cls(DataTypes.date, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def fit_beta(cls, draw_id):
+        return cls(DataTypes.fit_beta, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def parameter(cls, draw_id):
+        return cls(DataTypes.parameter, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    @classmethod
+    def regression_beta(cls, draw_id):
+        return cls(DataTypes.regression_beta, DRAW_FILE_TEMPLATE, draw_id=draw_id)
+
+    # factories related to forecasting
+    @classmethod
+    def beta_scales(cls, scenario, draw_id):
+        return cls(DataTypes.beta_scales, DRAW_FILE_TEMPLATE, scenario=scenario, draw_id=draw_id)
+
+    @classmethod
+    def components(cls, scenario, draw_id):
+        return cls(DataTypes.components, DRAW_FILE_TEMPLATE, scenario=scenario, draw_id=draw_id)
+
+    # other methods/properties
+    @property
+    def key(self):
+        return self.template.format(**self.key_args)
+
+    def nested_dirs(self):
+        """
+        Return any nesting structure required between the root save location
+        and the actual file.
+
+        For regression tasks this is just the data_type. For forecasting tasks
+        this is the scenario and the data type.
+        """
+        if 'scenario' in self.key_args:
+            return "/".join([self.key_args['scenario'], self.data_type])
+        else:
+            return self.data_type
+
+    @property
+    def seed(self):
+        """
+        Returns a seed value for enabling concurrency.
+
+        This is a sort of hack - we embed the concurrency story into the Keys
+        class which can then inform other things.
+        """
+        # everything is done by draw_id, so always use that as the seed value
+        return "draw-{draw_id}".format(**self.key_args)
+
+    def __repr__(self):
+        return f"Keys({self.data_type!r}, {self.template!r}, **{self.key_args!r})"
```

Comment:
I have not seen this syntax before, out of curiosity what does the !r do?

Reply:
`!r` substitutes `repr` output. `!s` substitutes `str` output (and is used by default)

```
>>> bar = "words"
>>> f"{bar}"
'words'
>>> f"{bar!s}"
'words'
>>> f"{bar!r}"
"'words'"
```

Reply:
ah very neat, thank you!